src/Microsoft.SqlTools.ServiceLayer/FileBrowser/FileBrowserService.cs

```diff
@@ -132,90 +132,137 @@ public void InitializeService(ServiceHost serviceHost)
-                }
-                else if (hasPendingQueueItems)
-                { 
-                    this.expandNodeQueue.StopQueueProcessor(DefaultExpandTimeout);
-                }
+                var task = Task.Run(() => RunFileBrowserCloseTask(fileBrowserParams, requestContext))
+                    .ContinueWithOnFaulted(null);
+                await requestContext.SendResult(new FileBrowserCloseResponse() { Succeeded = true });
+            }
+            catch
+            {
+                await requestContext.SendResult(new FileBrowserCloseResponse());
             }
-
-            await requestContext.SendResult(response);
         }
 
         #endregion
 
         public void Dispose()
         {
-            this.expandNodeQueue.Dispose();
+            this.fileBrowserQueue.Dispose();
         }
 
-        internal async Task RunFileBrowserOpenTask(FileBrowserOpenParams fileBrowserParams, RequestContext<bool> requestContext)
+        internal async Task RunFileBrowserCloseTask(FileBrowserCloseParams fileBrowserParams, RequestContext<FileBrowserCloseResponse> requestContext)
         {
-            FileBrowserOpenedParams result = new FileBrowserOpenedParams();
-            SqlConnection conn = null;
-            FileBrowserOperation browser = null;
-            bool isCancelRequested = false;
-
-            if (this.expandNodeQueue.IsCancelRequested)
-            {
-                this.expandNodeQueue.StartQueueProcessor();
-            }
-
+            FileBrowserCloseResponse result = new FileBrowserCloseResponse();
             try
             {
-                if (!fileBrowserParams.ChangeFilter)
+                FileBrowserOperation operation = null;
+                ConnectionInfo connInfo;
+                ownerToFileBrowserMap.TryGetValue(fileBrowserParams.OwnerUri, out operation);
+                this.ConnectionServiceInstance.TryFindConnection(fileBrowserParams.OwnerUri, out connInfo);
+
+                if (operation != null && connInfo != null)
                 {
-                    ConnectionInfo connInfo;
-                    this.ConnectionServiceInstance.TryFindConnection(fileBrowserParams.OwnerUri, out connInfo);
-                    if (connInfo != null)
+                    if (!operation.FileTreeCreated)
                     {
-                        // Open new connection for each Open request
-                        conn = ConnectionService.OpenSqlConnection(connInfo, this.serviceName);
-                        browser = new FileBrowserOperation(conn, fileBrowserParams.ExpandPath, fileBrowserParams.FileFilters);
+                        operation.Cancel();
                     }
-                }
-                else
-                {
-                    ownerToFileBrowserMap.TryGetValue(fileBrowserParams.OwnerUri, out browser);
-                    if (browser != null)
+
+                    // Clear queued items
+                    this.fileBrowserQueue.ClearQueuedItems();
```

Comment:
>ClearQueuedItems [](start = 42, length = 16)

so the design is we can only have 1 file browsing operation running at a time?  this is fine for Carbon use case, but in general we try to design features to work correctly in other tools.  For example would this work in a multi-user app where multiple file browses are going on at a time?

I think this design is probably fine for now, though please be sure to think through scenarios beyond carbon UI when designing features.

Reply:
Yes, for now since we can have only one file browser dialog at a time, I added all the operations in a queue. But this will need to be changed like in the scenario you mentioned so that multiple file browser operations can work. I'll create an issue to track this.